Mathlib/CategoryTheory/MorphismProperty/Presheaf.lean

```diff
@@ -0,0 +1,209 @@
+/-
+Copyright (c) 2024 Calle Sönne. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Calle Sönne, Joël Riou, Ravi Vakil
+-/
+
+import Mathlib.CategoryTheory.Limits.Shapes.Pullback.CommSq
+
+/-!
+
+# Representable morphisms of presheaves
+
+In this file we define and develop basic results on the representability of morphisms of presheaves.
+
+## Main definitions
+
+* `Presheaf.representable` is a `MorphismProperty` expressing the fact that a morphism `f : F ⟶ G`
+  of presheaves is representable, i.e. for any `g : yoneda.obj X ⟶ G`, there exists a pullback
+  square of the following form
+```
+  yoneda.obj Y --yoneda.map snd--> yoneda.obj X
+      |                                |
+     fst                               g
+      |                                |
+      v                                v
+      F ------------ f --------------> G
+```
+
+## API
+
+Given `hf : Presheaf.representable f`, with `f : F ⟶ G` and `g : yoneda.obj X ⟶ G`, we provide:
+* `hf.pullback g` which is the object in `C` such that `yoneda.obj (hf.pullback g)` forms a
+  pullback square of `f` and `g`.
+* `hf.snd g` is the morphism `hf.pullback g ⟶ X`
+* `hf.fst g` is the morphism `yoneda.obj (hf.pullback g) ⟶ F`
+*  Whenever `f` is of type `yoneda.obj Y ⟶ G`, we also have `hf.fst' g : hf.pullback g ⟶ Y`
+which is the preimage under `yoneda` of `hf.fst g`.
+* `hom_ext`, `hom_ext'`, `lift`, `lift'` are variants of the universal property of
+  `yoneda.obj (hf.pullback g)`, where as much as possible has been formulated internally to `C`.
+* `symmetry` and `symmetryIso` are variants of the fact that pullbacks are symmetric for
+  representable morphisms, formulated internally to `C`.
+
+-/
+
+
+namespace CategoryTheory
+
+open Category Limits
+
+universe v u
+
+variable {C : Type u} [Category.{v} C]
+
+/-- A morphism of presheaves `f : F ⟶ G` is representable if for any `X : C`, and any morphism
+`g : yoneda.obj X ⟶ G`, there exists a pullback square
+```
+yoneda.obj Y --yoneda.map snd--> yoneda.obj X
+    |                                |
+   fst                               g
+    |                                |
+    v                                v
+    F ------------ f --------------> G
+``` -/
+def Presheaf.representable : MorphismProperty (Cᵒᵖ ⥤ Type v) :=
+  fun F G f ↦ ∀ ⦃X : C⦄ (g : yoneda.obj X ⟶ G), ∃ (Y : C) (snd : Y ⟶ X)
+    (fst : yoneda.obj Y ⟶ F), IsPullback fst (yoneda.map snd) f g
```

Comment:
I wonder if this should be generalized, i.e. let `Cᵒᵖ ⥤ Type v` just be any category, and `yoneda` just be any functor into it. The definition makes sense, and also a reasonable amount of the API below.

Reply:
(And then have a specialized version for `yoneda`.)

Reply:
`hasRelativePullbacks`, perhaps?

Reply:
(I don't intend this comment to block merging, if it doesn't sounds like a good idea!)

Reply:
I think this sound interesting / fun, so I will try it

Reply:
Actually, maybe it's not so bad.

Reply:
Done!